### ISSUE TITLE
new Dict(generator) inferred type doesn't match return type

### ISSUE THREAD

**Issue by gaika** [JuliaLang/julia#19243]
``Dict(generator)`` inferred type doesn't match return type:
```
julia> Test.@inferred Dict(a => b for (a,b) in [(1,2)])
ERROR: return type Dict{Int64,Int64} does not match inferred return type Union{Dict{Int64,Int64},Dict{Union{},Union{}}}
 in error(::String) at ./error.jl:21
```

But everything was fine when using deprecated syntax with array of pairs:

```
julia> Test.@inferred Dict([a => b for (a,b) in [(1,2)]])

WARNING: deprecated syntax "[a=>b for (a,b) in c]".
Use "Dict(a=>b for (a,b) in c)" instead.
Dict{Int64,Int64} with 1 entry:
  1 => 2
```

Results in ~50% slowdown in heavy use of that dictionary.

VERSION
v"0.6.0-dev.1146"

**Comment by yuyichao** [comment]
The second `@inferred` is not testing anything at all.

**Comment by gaika** [comment]
Ah, sorry, you are right, but still, constructor difference is  causing a huge slowdown in my code:
```
function f()
    d1 = Dict(1 => 2)
    d2 = Dict(t => t for t in 1)
end
@code_warntype f()
Variables:
  #self#::#f
  d1::Dict{Int64,Int64}
  d2::Union{Dict{Int64,Int64},Dict{Union{},Union{}}}
```

**Comment by JeffBezanson** [comment]
@gaika As a temporary workaround you can request a more specific type of dictionary; `Dict{T,S}(...)`.

**Comment by JeffBezanson** [comment]
Closed by #19245
